ResourceStatusSystem/Agent/SiteInspectorAgent.py

```diff
@@ -0,0 +1,183 @@
+      return sitesToBeChecked
+    self.sitesToBeChecked = sitesToBeChecked[ 'Value' ]
+
+    queueSize   = self.sitesToBeChecked.qsize()
+    pollingTime = self.am_getPollingTime()
+
+    # Assigns number of threads on the fly such that we exhaust the PollingTime
+    # without having to spawn too many threads. We assume 10 seconds per element
+    # to be processed ( actually, it takes something like 1 sec per element ):
+    # numberOfThreads = elements * 10(s/element) / pollingTime
+    numberOfThreads = int( math.ceil( queueSize * 10. / pollingTime ) )
+
+    self.log.info( 'Needed %d threads to process %d elements' % ( numberOfThreads, queueSize ) )
+
+    for _x in xrange( numberOfThreads ):
+      jobUp = self.threadPool.generateJobAndQueueIt( self._execute )
+      if not jobUp[ 'OK' ]:
+        self.log.error( jobUp[ 'Message' ] )
+
+    self.log.info( 'blocking until all elements have been processed' )
+    # block until all tasks are done
+    self.sitesToBeChecked.join()
+    self.log.info( 'done')
+
+    return S_OK()
+
+
+  def getSitesToBeChecked( self ):
+    """ getElementsToBeChecked
+
+    This method gets all the rows in the <self.elementType>Status table, and then
+    discards entries with TokenOwner != rs_svc. On top of that, there are check
+    frequencies that are applied: depending on the current status of the element,
+    they will be checked more or less often.
+
+    """
+
+    toBeChecked = Queue.Queue()
+
+    sites = self.siteClient.getSites('All')
+    if not sites[ 'OK' ]:
+      return sites
+
+    # filter elements by Type
+    for site in sites[ 'Value' ]:
+
+      # get the current status
+      status = self.siteClient.getSiteStatuses( [site] )
+      if not status['OK']:
+        return status
+
+      if not status['Value'][site]:
+        status = 'Unknown'
+      else:
+        status = status['Value'][site]
+
+      # We add lowerElementDict to the queue
+      toBeChecked.put( { 'status': status, 'name': site, 'site' : site, 'element' : 'Site', 'statusType': 'all', 'elementType': 'Site' } )
+
+    return S_OK( toBeChecked )
+
+
+  # Private methods ............................................................
+
+  def _execute( self ):
+    """
+      Method run by the thread pool. It enters a loop until there are no elements
+      on the queue. On each iteration, it evaluates the policies for such element
+      and enforces the necessary actions. If there are no more elements in the
+      queue, the loop is finished.
+    """
+
+    pep = PEP( clients = self.clients )
+
+    while True:
+
+      try:
+        site = self.sitesToBeChecked.get_nowait()
+      except Queue.Empty:
+        return S_OK()
+
+      resEnforce = pep.enforce( site )
+      if not resEnforce[ 'OK' ]:
+        self.log.error( 'Failed policy enforcement', resEnforce[ 'Message' ] )
+        self.sitesToBeChecked.task_done()
+        continue
+
+      # Used together with join !
+      self.sitesToBeChecked.task_done()
+
```

Comment:
All this file is mostly a copy/paste from the `ElementInspectorAgent`, I was wondering if it wouldn't be better to have a "InspectorAgent" that is generic, or alternatively create specialize an agent by inheriting from it (for re-usable code).